### PR TITLE
[Embedding] remove kernel functions from operator

### LINKED ISSUE #443: Kernel refactor for `Embedding`

### PR DESCRIPTION
**Pull request by rajas1998**
**Description of changes:**
This PR removes kernel functions from embedding.h and embedding.cc into separate embedding_kernels.h and embedding_kernels.cu. It also modifies embedding.cpp accordingly and renames the file to embedding_kernels.cpp.

**Related Issues:**

Linked Issues:
- Issue #303

Issues closed by this PR:
- Closes #443

**Before merging:**

- [ ] Did you update the [flexflow-third-party](https://github.com/flexflow/flexflow-third-party) repo, if modifying any of the Cmake files, the build configs, or the submodules? No

### CODE REVIEW

File: src/ops/kernels/embedding_kernels.cpp [hunk 1]
```diff
@@ -27,6 +27,220 @@ using Legion::Rect;
 using Legion::Runtime;
 using Legion::Task;
 
+namespace Kernels {
+namespace Embedding {
+
+/*static*/
```
Comment (virena) [inline src/ops/kernels/embedding_kernels.cpp:33]:
```suggestion
```

File: src/ops/kernels/embedding_kernels.cpp [hunk 2]
```diff
@@ -388,177 +510,53 @@ void Embedding::backward_kernel(TI const *input_ptr,
...
-                     GET_BLOCKS(size),
-                     CUDA_NUM_THREADS,
-                     0,
-                     stream,
-                     ptr,
-                     size,
-                     p);
-}
-
 #ifdef DEADCODE
```
Comment (virena) [inline src/ops/kernels/embedding_kernels.cpp:520]: Can this deadcode be safely removed?

**Comment by virena** [comment]
Why is embedding only handled in `fused.cu` and not in `fused.cpp`?

File: src/ops/kernels/embedding_kernels.cpp (hunk 2 above)
Comment (rajas1998) [reply src/ops/kernels/embedding_kernels.cpp:520]: Not sure how this deadcode was being used. In the current flow its not being used, kept it just to maintain backward compatibility.

**Comment by rajas1998** [comment]
Embedding was only handled in fused.cu and not fused.cpp even before making the changes for the kernels. I am not sure why embedding has not been added into fused.cpp - might be some legacy reason which I am not sure about

**Comment by virena** [comment]
> Embedding was only handled in fused.cu and not fused.cpp even before making the changes for the kernels. I am not sure why embedding has not been added into fused.cpp - might be some legacy reason which I am not sure about

@lockshaw do we want to address this here or later?

**Review by virena — APPROVED** [review]

**Comment by lockshaw** [comment]
> > Embedding was only handled in fused.cu and not fused.cpp even before making the changes for the kernels. I am not sure why embedding has not been added into fused.cpp - might be some legacy reason which I am not sure about
> 
> @lockshaw do we want to address this here or later?

I'm actually not sure why we have both `fused.cpp` _and_ `fused.cu` at all, as the code looks pretty platform agnostic. @virena conditionally assigned to you, feel free to reassign if there's someone better to work on it #569

File: src/ops/kernels/embedding_kernels.cpp (hunk 2 above)
Comment (lockshaw) [reply src/ops/kernels/embedding_kernels.cpp:519]: #570

**Review by lockshaw — APPROVED** [review]

**Pull request merged**